Add re_exports for libraries

rgrinberg opened the pull request:
This is an attempt to implement a proper solution described by @avsm in #2595 

There are some things that need to be discussed before I finish the implementation:

* `re_exports` should only allow libraries already present in `libraries`.

* When we write `re_exports` to the the `dune-package` files, should we just write the closure of re-exported libraries? This is simpler to implement, but perhaps it makes sense to copy what we do for requires? I'm not a fan of that as it complicates things and I don't see the benefit. What was the reason we got rid of writing the transitive closure in the requires field?

diml commented:
> re_exports should only allow libraries already present in libraries.

That seems right

> When we write re_exports to the the dune-package files, should we just write the closure of re-exported libraries? This is simpler to implement, but perhaps it makes sense to copy what we do for requires? I'm not a fan of that as it complicates things and I don't see the benefit. What was the reason we got rid of writing the transitive closure in the requires field?

We don't want to re-export the transitive dependencies though

rgrinberg commented:
Just to clarify, this is the behavior that I expect. If a library `a` re-exports `b` and the library `b` re-exports `c`, I would expect that users of `a` will find both `b` and `c` visible. @diml do you agree with this behavior?

diml commented:
Yep, I agree with this. What I had in mind is: if c depends on d but doesn't re-export d, then users of a shouldn't see d.

rgrinberg commented:
Okay, good. So my question was for how should we output the dune-package files for the libraries. For `a` re-exports `b` which re-exports `c`, we can have `re_exports` for `a` be:

* Non transitive and just have `(re_exports b)`. Dune will walk the transitive re-exports fields to get a full list of re-exports

* Transitive and have the full list of re-exports. In this case it would be `(re_exports b c)`.

I think the latter is easier to implement but we veered away from it for normal dependencies.

diml commented:
Putting the transitive list seems good. IIRC, for normal dependencies there was an issue when depending on non-dune libraries that were using predicates to select different dependencies. If you store the transitive dependencies then the choice is written in stone once and for all and you can't change it afterwards. In this case, it's not a concern though.

And re-exporting things should clearly be a "sticky" property.

rgrinberg committed:
Add tests for re-exporting transitive dependencies

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
The distinction between public and private names is complicating the
implementation of this issue a bit. I'm not sure whether the predicate language
should apply to private or public names.

I see a couple of options:

* As anywhere else in dune, public and private names should be interchangeable. So
the predicate language should apply to both names.

* Another good option is just to allow filtering based on the names specified by
the user. So if a user uses the name `bar.foo` in the `libraries` field, it
would have to be specified as `bar.foo` in the `re_exports` field as well.

rgrinberg commented:
After discussing this with Jeremie, the conclusion is just that it's simpler to avoid these altogether by getting rid of the predicate language in this case. We can make do with library names. I will update this PR accordingly.

rgrinberg committed:
Add test for re exporting an invalid library

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented:
That seems fine yes. It means that we can't re-export a dependency under a `(select ...)`, but maybe that's a good thing.

rgrinberg commented:
> That seems fine yes. It means that we can't re-export a dependency under a (select ...), but maybe that's a good thing.

I think this needs explicit support from the old scheme as well. So either scheme will require more work to make selectable exports work

diml commented:
Yh, plus it's not even clear that it's need. In general, one dispatch between different dependencies for the implementation, not the interface. So a constructor rather than a field seems fine.

rgrinberg committed:
Add re_exports field to libraries

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Implement re_exports dependencies

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Use re_export as a constructor

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Move Lib_dep into own module

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Remove Dune_package.Lib.requires

it wasn't used anywhere but the tests

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add support for re_exports constructor

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
I've now implemented the `re_exports` constructors. The set of changes that this required turned out to be much larger than I expected. Some comments why this might be the case:

* The distinction between `Lib_info.Deps.{Simple,Complex}` doesn't seem very helpful. `Simple` can always be injected to `Complex` and it doesn't really add any type safety either.

* To increase the type safety, it could perhaps be useful to parameterize `Lib_info.t` over the kind of `requires` that it contains. The idea being that there are basically 2 modes: evaluated vs. unevaluated. This property is orthogonal to local vs. external as it's a matter of whether we obtained the info from the Dune_file.Library.t (unevaluated), Lib.t (evaluated), Dune_package, (evaluated).

I also removed `Dune_package.Lib.requires` as it was only used in the tests.

rgrinberg committed:
Add CHANGES

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Do not allow re_export in executables or cinaps

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Improve error message in dep spec

Keywords apart from select and re_export should give a proper error
message.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented:
How would you inject `Simple` into `Complex`? One reason to keep them distinct is that the interpretation of complex dependencies is much more complex and costly than simple ones, while simple ones are the most common case by far.

rgrinberg commented:
> How would you inject Simple into Complex? One reason to keep them distinct is that the interpretation of complex dependencies is much more complex and costly than simple ones, while simple ones are the most common case by far.

I would inject it with something like `Complex (List.map ~f:Lib_dep.direct simple)`.

Also, I wasn't aware that Complex was so much more expensive than Simple. When I looked, `resolve_complex_deps` looked well behaved when there were no select dependencies.

jeremiedimino committed:
Update changelog

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

diml commented:
Ah indeed, I was confusing with injecting direct dependencies into select ones. In that case, yes that seems fine to me.

diml commented on test/blackbox-tests/test-cases/re-exported-deps/run.t:
```diff
@@ -0,0 +1,45 @@
+
+Attempting to re-export dependencies outside of libraries fails:
+  $ dune build --root re-export-bad-attempt @all
+  Entering directory 're-export-bad-attempt'
```
Shouldn't that fail?

rgrinberg replied:
I guess not, because `(re_export foo)` will always work as long as `foo` works. It's interpreted as a modifier on the library name, so it's no longer possible to have mismatches like we did before.

diml replied:
Should the comment be updated then? The comment makes me think that the command is expected to fail

diml commented:
> I also removed Dune_package.Lib.requires as it was only used in the tests.

Acutally hang on, I'm needing this in #2528

rgrinberg commented:
> Acutally hang on, I'm needing this in #2528

You can't use Lib_info.requires?

diml commented:
let me try

diml commented:
I do need the evaluated version but I ended up doing it another way, so I don't need it in the end

rgrinberg commented:
Indeed. I think I will remove this test as I’m not sure how it could make
sense with the new constructors

On Wed, Sep 18, 2019 at 12:46 AM Jérémie Dimino <notifications@github.com>
wrote:

> *@diml* commented on this pull request.
> ------------------------------
>
> In test/blackbox-tests/test-cases/re-exported-deps/run.t
> <https://github.com/ocaml/dune/pull/2605#discussion_r325246284>:
>
> > +   (main_module_name Bbb)
> +   (modes byte native)
> +   (modules (singleton (name Bbb) (obj_name bbb) (visibility public) (impl))))
> +  (library
> +   (name pkg.ccc)
> +   (kind normal)
> +   (archives (byte ccc/ccc.cma) (native ccc/ccc.cmxa))
> +   (plugins (byte ccc/ccc.cma) (native ccc/ccc.cmxs))
> +   (foreign_archives (native ccc/ccc$ext_lib))
> +   (main_module_name Ccc)
> +   (modes byte native)
> +   (modules (singleton (name Ccc) (obj_name ccc) (visibility public) (impl))))
> +
> +Attempting to re-export dependencies outside of libraries fails:
> +  $ dune build --root re-export-bad-attempt @all
> +  Entering directory 're-export-bad-attempt'
>
> Should the comment be updated then? The comment makes me think that the
> command is expected to fail
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/dune/pull/2605?email_source=notifications&email_token=AABB56ZMUAXSQQIJDOUHHPLQKD3W7A5CNFSM4ITGFG2KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCE7V4YY#discussion_r325246284>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABB5627NLBCUTSFH7JYCUTQKD3W7ANCNFSM4ITGFG2A>
> .
>

rgrinberg committed:
Add test for re-exporting deps in executables

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Remove irrelevant test

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Remove "Simple" deps

They are just a performance optimization that isn't buying us all that much

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Move select resolution to own function

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Simplify select resolutionn

Cons the accumulated library list rather than appending.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented on src/dune/lib.ml:
```diff
@@ -1871,10 +1874,16 @@ let to_dune_lib ({ info; _ } as lib) ~modules ~foreign_objects ~dir =
+    List.map ~f:Lib_dep.direct requires
+    @ List.map ~f:Lib_dep.re_export re_exports
+  in
+  let info = Lib_info.set_requires info requires in
```
These successive `Lib_info.set_xxx` feels inefficient BTW, we should have a single `Lib.set` function that takes multiple fields to avoid creating many intermediate records.

rgrinberg replied:
Yeah, good call. I will turn this into a single call.

diml commented on src/dune/lib.ml:
```diff
@@ -1871,10 +1874,16 @@ let to_dune_lib ({ info; _ } as lib) ~modules ~foreign_objects ~dir =
-  Dune_package.Lib.make ~info ~requires ~modules:(Some modules)
-    ~main_module_name
+  let+ re_exports = lib.re_exports in
+  let re_exports = List.map ~f:(fun t -> (Loc.none, t.name)) re_exports in
```
Why not `add_loc re_exports`?

jeremiedimino committed:
Decoder.file -> Decoder.filename + doc

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

diml commented on src/dune/lib.ml:
```diff
@@ -1118,80 +1123,78 @@ end = struct
+              let+ acc_re_exports = acc_re_exports in
+              lib :: acc_re_exports
+            in
+            (acc_res, acc_selects, acc_re_exports)
```
Why don't we add `lib` to `acc_res`? IIIUC, re-exported libraries are not part of `t.requires`. That seems odd to me, in particular it seems that they don't end up in `Compile.direct_requires`.

rgrinberg replied:
I can do that, but I think we still need to maintain a separate list of re_exports. We need to walk that list to get a list of all re_exported libraries.

rgrinberg replied:
Actually, they do end up in direct_requires. `resolve_user_deps` resolves all re_exported dependencies.

diml replied:
Yeah, a separate re-exports list seems fine. Basically I was just finding it odd that `re_exports` wasn't included in `res` given that each re-exported dependency is also a plan dependency.

diml replied:
For instance, dependencies coming from `select` forms are also included in `res`

diml commented on src/dune/lib.ml:
```diff
@@ -1871,10 +1874,16 @@ let to_dune_lib ({ info; _ } as lib) ~modules ~foreign_objects ~dir =
+  let re_exports = List.map ~f:(fun t -> (Loc.none, t.name)) re_exports in
+  let requires =
+    List.map ~f:Lib_dep.direct requires
+    @ List.map ~f:Lib_dep.re_export re_exports
```
This is related to the comment about `acc_res`. I feel like this `requires` should already include re-exported libraries

rgrinberg replied:
But `requires` is just a list of libraries. We don't know which library was marked with `re_export`, so we can't output a correct list for dune-package.

rgrinberg committed:
Use add_loc prepare re_exports list

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Update Lib_info all at once

Instead of setting it field by field, we just update everything in a
single call before getting it into a dune_package ready form.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add re_exports to requires

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
By the way, a related change that we should consider is to change all these library lists to `Loc.t Lib.Map.t`. There are some situations when the order of the list is arbitrary, and we aren't being careful about making sure it has not duplicates. If it's really so important to preserve this order (Unless, it's for preprocessors, I don't really see a reason), we should use an abstract list that makes sure no duplicates are added.

diml commented:
Hmm, I feel a bit nervous about changing the order. In some cases, the order actually matters for some weird reason :/ It doesn't have to be specified, but changing it might introduce problems.

rgrinberg committed:
Use let+/and+

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>
